hal/src/stm32f2xx/concurrent_hal.cpp

```diff
@@ -489,3 +496,47 @@ int os_timer_is_active(os_timer_t timer, void* reserved)
 {
     return xTimerIsTimerActive(timer) != pdFALSE;
 }
+
+static AtomicFlagMutex<os_result_t, os_thread_yield> flash_lock;
+void __flash_acquire() {
+    if (HAL_IsISR()) {
+        PANIC(UsageFault, "Flash operation from IRQ");
+    }
+    flash_lock.lock();
+}
+
+void __flash_release() {
+    flash_lock.unlock();
+}
+
+#define DCT_LOCK_TIMEOUT 10000
+
+static RecursiveMutex dctLock(DCT_LOCK_TIMEOUT);
```

Comment:
Seem useful at least for debugging. There is an assert that checks whether the lock was successfully acquired or not.